server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -2189,6 +2189,53 @@ protected void validateConditionsToReplaceDiskOfferingOfVolume(VolumeVO volume,
         s_logger.info(String.format("Changing disk offering to [uuid=%s] while migrating volume [uuid=%s, name=%s].", newDiskOffering.getUuid(), volume.getUuid(), volume.getName()));
     }
 
+    /**
+     *  Checks if the target storage supports the new disk offering.
+     *  This validation is consistent with the mechanism used to select a storage pool to deploy a volume when a virtual machine is deployed or when a new data disk is allocated.
+     *
+     *  The scenarios when this method returns true or false is presented in the following table.
+     *
+     *   <table border="1">
+     *      <tr>
+     *          <th>#</th><th>Disk offering tags</th><th>Storage tags</th><th>Does the storage support the disk offering?</th>
+     *      </tr>
+     *      <body>
+     *      <tr>
+     *          <td>1</td><td>A,B</td><td>A</td><td>NO</td>
+     *      </tr>
+     *      <tr>
+     *          <td>2</td><td>A,B,C</td><td>A,B,C,D,X</td><td>YES</td>
+     *      </tr>
+     *      <tr>
+     *          <td>3</td><td>A,B,C</td><td>X,Y,Z</td><td>NO</td>
+     *      </tr>
+     *      <tr>
+     *          <td>4</td><td>null</td><td>A,S,D</td><td>YES</td>
+     *      </tr>
+     *      <tr>
+     *          <td>5</td><td>A</td><td>null</td><td>NO</td>
+     *      </tr>
+     *      <tr>
+     *          <td>6</td><td>null</td><td>null</td><td>YES</td>
+     *      </tr>
+     *      </body>
+     *   </table>
+     */
+    protected boolean doesTargetStorageSupportNewDiskOffering(StoragePool destPool, DiskOfferingVO newDiskOffering) {
```

Comment:
Well, I am not sure if we have some other places that can use this method. I would normally only set a method to public and declare it in an interface if I need really need it. I can declare it if you think it is important though. 

P.S.: I find our interface usage sometimes quite silly with single implementations.  A good example is using interface for POJOs. It does not provide benefits, the only thing it does is adding dozens more lines for every object.

Reply:
ok, nice subject: another valid usage is mutual dependent services. Or dependency injection.
I this case my concern is that the implementation of tag matching is doen in several places while it should be uniform and thus implemented centrally. doesTargetStorageASupportNewDiskOffering() is a specialisation of does_this_entities_tags_match_the_resource_it_is_going_to(entity, resource) which is applicable to volume/storage in case of new offering but also in case of migration, and is also applicable for vm-serviceoffering/host-cluster-pod etc. I cant' think of any others by head but these are surely there as well.

Reply:
This new method that I implemented is the consolidation of the logic of tag matching in ACS, which is not extracted in a specific method and spread around the code. That is one of the reasons why I “made the mistake” in one of the features I developed where the tag matching was being too strict (requiring all tags to match) and was not making sense to operation guys.

To answer your question, yes this method can be used in other places. I am not sure if simply declaring the method in the interface will help much though. The person that is willing to consolidate these execution flow could do both, declare this method in the interface, and then use it in all of those places that you mentioned.

What do you think?

Reply:
I think it is a nice future plan to extract the logic to a tagging utility and then make this method use it and be given a more general name like
```
 doesTargetStorageSupportDiskOffering() // not no 'New' in there
```
and have it call the utility
```
 matchOk(Set<String> tags, List<String> supported);
```
plan?

Reply:
matchOk is just a bad method-name i came up with. we need a set of methods that do the matching for us for
Set<String> tags, List<String> supported or
List<List<String>> tags, List<String> supported or
String tag(s), List<String> supported etc

Reply:
Ok, but in this case I think we should create methods on demand. I mean, we do not know what we need yet. Therefore, there is no reason to start coding methods that we do not even know if we will use them.

Reply:
ok, maybe postpone this discussion to a beer session ;)

Reply:
of course I meant hackathon

Reply:
Ahaha.. I do not know why, but I would prefer a beer session...

I will do the change you asked later (making the method public in the interface).